doc/decisions/lookup_every_key.md

```diff
@@ -0,0 +1,47 @@
+# Lookup Every Key
+
+## Problem
+
+On structures like maps or [arrays](array.md) there are different
+possibilities which keys are looked up in the KeySet and which
+are simply iterated.
+
+Without any guidelines, applications would provide arbitrary inconsistent
+behavior.
+
+## Constraints
+
+## Assumptions
+
+- Applications that have good reasons to ignore the guidelines
+  (e.g. they only read from one namespace), are allowed to do so.
+
+## Considered Alternatives
+
+- only lookup the roots and then iterate over the next keys
+- let the applications do what they want without any guideline
+
+## Decision
+
+Every key that an application wants to use, should be looked up
+with `ksLookup`.
```

Comment:
I agree with this decision, but I would go a step further:

- Applications that use the KDB as a config provider, MUST lookup every key with `ksLookup` using a cascading lookup key. These applications MUST NOT access a namespaced key directly.
- Applications whose primary purpose is to manipulate the KDB (e.g. `kdb`), are exempt from the rule and may do whatever they need. This includes accessing and modifying individual namespaces directly.

This would allow further optimizations in the future. We could add an option to `kdbGet` that transforms the `KeySet` in such a way that only the actually accessible `Key`s will be returned.

Reply:
I made it a bit stricter 83d0eb15fe03abc0f26f9ba4900c8afb302a6b7e but I would not stretch it too far. I think there are some use-cases where namespaced lookups are appropriate in system tools, e.g. working with configs equivalent to /etc/passwd.

>   This would allow further optimizations in the future.

Yes :+1:

Reply:
Direct lookups for `system:/` keys can make sense, although it might be more appropriate to have a `spec:/` that says "this key can only exist in `system:/`". Then you would also get an error, if a `user:/` key for something like `/etc/passwd` exists, instead of the key just being ignored.